cli/azd/pkg/tools/github/github.go

```diff
@@ -323,3 +377,184 @@ func isEnvVarTokenSource(message string) bool {
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+	return extractedAt, nil
+}
+func extractFromTar(src, dst string) (string, error) {
+	gzFile, err := os.Open(src)
+	if err != nil {
+		return "", err
+	}
+	defer gzFile.Close()
+
+	gzReader, err := gzip.NewReader(gzFile)
+	if err != nil {
+		return "", err
+	}
+	defer gzReader.Close()
+
+	var extractedAt string
+	// tarReader doesn't need to be closed as it is closed by the gz reader
+	tarReader := tar.NewReader(gzReader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find gh cli within tar file: %w", err)
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "gh" {
+			filePath := filepath.Join(dst, fileName)
+			ghCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer ghCliFile.Close()
+			_, err = io.Copy(ghCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+
+	return extractedAt, nil
+}
+
+// extractGhCli gets the Github cli from either a zip or a tar.gz
+func extractGhCli(src, dst string) (string, error) {
+	if strings.Contains(src, ".zip") {
+		return extractFromZip(src, dst)
+	} else if strings.Contains(src, ".tar.gz") {
+		return extractFromTar(src, dst)
+	}
+
+	// exe rights for file
+	// if err := f.Chmod(osutil.PermissionExecutableFile); err != nil {
+	// 	return err
+	// }
+
+	return "nil", fmt.Errorf("Unknown format")
+}
+
+// GetGitHubCliImplementation defines the contract function to acquire the GitHub cli.
+// The `outputPath` is the destination where the github cli is place it.
+type GetGitHubCliImplementation func(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	outputPath string) error
+
+// ExtractGitHubCliFromFileImplementation defines how the cli is extracted
+type ExtractGitHubCliFromFileImplementation func(src, dst string) (string, error)
+
+// downloadGh downloads a given version of GitHub cli from the release site.
+func downloadGh(
+	ctx context.Context,
+	transporter policy.Transporter,
+	ghVersion semver.Version,
+	extractImplementation ExtractGitHubCliFromFileImplementation,
+	path string) error {
+
+	// arm and x86 not supported (similar to bicep)
+	var releaseName string
+	switch runtime.GOOS {
+	case "windows":
+		releaseName = "gh_2.22.1_windows_amd64.zip"
+	case "darwin":
+		releaseName = "gh_2.22.1_macOS_amd64.tar.gz"
+	case "linux":
+		releaseName = "gh_2.22.1_linux_amd64.tar.gz"
+	default:
+		return fmt.Errorf("unsupported platform")
```

Comment:
We should add ARM64 support. We don't want M1 users to be left in a bad place.

Reply:
do we have ARM in CI already?  
Usually I wouldn't add something which we can't test on CI or in manual test flows

Reply:
@danieljurek What did we discover as part of the M1 work item?

Reply:
Note:  it is the same story for `bicep`
See release options:
https://github.com/Azure/bicep/releases

See what azd does:
![image](https://user-images.githubusercontent.com/24213737/217638703-11625276-7779-428d-9564-45bf69ac7ace.png)

So, I'd would leave arm for later, and make the change for both bicep and gh-cli.   
I don't know if we are currently shipping an ARM azd version, anyway

Reply:
Nope:

![image](https://user-images.githubusercontent.com/24213737/217639080-7e46d234-1dbd-49e1-809d-9752fb72e775.png)

azd does not run in ARM

Reply:
> @danieljurek What did we discover as part of the M1 work item?

That work was around enabling ARM64 on Linux (in devcontainer scenarios) and we didn't dive too deep into Apple Silicon scenarios specifically. 

In the case macOS the gh CLI publishes only one architecture: AMD64 (that is, Intel, not Apple Silicon). 

Without directly verifying, my expectation is that the gh CLI relies on Rosetta 2 to handle execution or, if [installed via homebrew, builds the gh CLI from source](https://github.com/Homebrew/homebrew-core/blob/34b732b38991b8b2f26b592d87287ebc77937f19/Formula/gh.rb#L16-L18)

Reply:
> @danieljurek What did we discover as part of the M1 work item?

That work was around enabling ARM64 on Linux (in devcontainer scenarios) and we didn't dive too deep into Apple Silicon scenarios specifically. 

In the case macOS the gh CLI publishes only one architecture: AMD64 (that is, Intel, not Apple Silicon). 

Without directly verifying, my expectation is that the gh CLI relies on Rosetta 2 to handle execution or, if [installed via homebrew, builds the gh CLI from source](https://github.com/Homebrew/homebrew-core/blob/34b732b38991b8b2f26b592d87287ebc77937f19/Formula/gh.rb#L16-L18)

Reply:
I will add arm for linux in a future PR.
And for macOS, we'll start an offline conversation with GitHub cli to know if they will support M1 and how.  For now, it won't be supported by azd

Reply:
@danieljurek Does this align with your expectations and/or plans for M1 support? Let's make sure we track the work somewhere.

Reply:
it should be part of https://github.com/Azure/azure-dev/issues/1314